src/parser/parser.cpp

```diff
@@ -448,6 +448,64 @@ std::vector<DatatypeType> Parser::mkMutualDatatypeTypes(
   }
 }
 
+Type Parser::mkFlatFunctionType(std::vector<Type>& sorts,
+                                Type range,
+                                std::vector<Expr>& flattenVars)
+{
+  if (range.isFunction())
+  {
+    std::vector<Type> domainTypes =
+        (static_cast<FunctionType>(range)).getArgTypes();
+    for (unsigned i = 0; i < domainTypes.size(); i++)
+    {
+      sorts.push_back(domainTypes[i]);
+      // the introduced variable is internal (not parsable)
+      std::stringstream ss;
+      ss << "__flatten_var_" << i;
+      Expr v = d_exprManager->mkBoundVar(ss.str(), domainTypes[i]);
+      flattenVars.push_back(v);
+    }
+    range = static_cast<FunctionType>(range).getRangeType();
+  }
+  if (sorts.empty())
+  {
+    return range;
+  }
+  else
+  {
+    return d_exprManager->mkFunctionType(sorts, range);
+  }
+}
+
+Type Parser::mkFlatFunctionType(std::vector<Type>& sorts, Type range)
+{
+  if (sorts.empty())
+  {
+    // no difference
+    return range;
+  }
+  else
+  {
+    while (range.isFunction())
+    {
+      std::vector<Type> domainTypes =
+          static_cast<FunctionType>(range).getArgTypes();
+      sorts.insert(sorts.end(), domainTypes.begin(), domainTypes.end());
+      range = static_cast<FunctionType>(range).getRangeType();
+    }
+    return d_exprManager->mkFunctionType(sorts, range);
+  }
+}
+
+Expr Parser::mkHoApply(Expr expr, std::vector<Expr>& args, unsigned startIndex)
+{
+  for (unsigned i = startIndex; i < args.size(); i++)
```

Comment:
for-each

Reply:
need for here because of startIndex, I'll make it size cached though

Reply:
need for here because of startIndex, I'll make it size cached though